src/transformers/models/llama/modeling_llama.py

```diff
@@ -889,8 +906,11 @@ def forward(
             raise ValueError("You have to specify either input_ids or inputs_embeds")
 
         past_key_values_length = 0
-        if past_key_values is not None:
-            past_key_values_length = past_key_values[0][0].shape[2]
+        if use_cache:
+            use_legacy_cache = not isinstance(past_key_values, Cache)
+            if use_legacy_cache:
+                past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+            past_key_values_length = past_key_values.get_seq_length()
```

Comment:
```suggestion
        if use_cache:
            use_legacy_cache = not isinstance(past_key_values, Cache) and past_key_values is not None
            if use_legacy_cache:
                past_key_values = DynamicCache.from_legacy_cache(past_key_values)
            elif past_key_values is None:
                past_key_values = DynamicCache()
            past_key_values_length = past_key_values.get_seq_length()
```

This has probably been discussed already but here I am: here in the first forward pass when using `generate`, we have `past_key_values` being None, and `use_legacy_cache` evaluates to `True`. This results in `from_legacy_cache` and `to_legacy_cache` being called at each decoding iteration. Is this intended for now?

Alternatively `DynamicCache()` may be initialized in `generation/utils.py` in case we care only about `generate` using the new cache implementation.

it's not clear to me why, but the proposed suggestion I give actually makes generation slower. Any idea what's happening?